### ISSUE TITLE
Test flake: TestReconcileInstanceWithParameters

### ISSUE THREAD

**Issue by pmorie** [kubernetes-incubator/service-catalog#472]
As demonstrated in https://travis-ci.org/kubernetes-incubator/service-catalog/builds/207242654

**Comment by vaikas-google** [comment]
I'm having a hard time deciphering which logs go with that test failure, but if it's the one right above the test:
E0302 19:27:10.788331     500 controller.go:334] Instance test-ns/test-instance references a non-existent ServiceClass test-serviceclass
--- FAIL: TestReconcileInstanceWithParameters (0.00s)
    controller_test.go:421: Unexpected condition status: expected True, got False

Then the lister is returning something wonky:
    serviceClass, err := c.serviceClassLister.Get(instance.Spec.ServiceClassName)
cause that's what's failing

[5:10]  
sure, but there's really no reason for it from what I can see here. The listers get set up the same way with or without the params

mbauer [5:13 PM] 
well, glog output will happen before the FAIL, and test log output will happen after.

[5:13]  
test log output will be indented, glog output will not be

mbauer [5:36 PM] 
TestReconcileInstanceWithParameters does seem flaky

[5:37]  
failed 1/10 times with `go test -v -run TestReconcileInstanceWithParameters ./pkg/controller/wip/ -race -count 10`

mbauer [5:42 PM] 
TestReconcileBrokerWithReconcileError has a race according to the race detector.

mbauer [5:48 PM] 
1000 runs
```rg FAIL testoutput
341:--- FAIL: TestReconcileBrokerWithReconcileError (0.01s)
23199:--- FAIL: TestReconcileBrokerWithReconcileError (0.01s)
43433:--- FAIL: TestReconcileBrokerWithReconcileError (0.01s)
79520:--- FAIL: TestReconcileBrokerWithReconcileError (0.01s)
86352:--- FAIL: TestReconcileBindingWithParameters (0.00s)
86732:--- FAIL: TestReconcileBrokerWithReconcileError (0.01s)
106969:--- FAIL: TestReconcileInstance (0.01s)
112198:--- FAIL: TestReconcileInstanceNonExistentServicePlan (0.01s)
112460:--- FAIL: TestReconcileInstance (0.00s)
112717:--- FAIL: TestReconcileBindingNonExistingServiceClass (0.01s)
```
(edited)

[5:48]  
Hmn, there's a thought.

mbauer [5:54 PM] 
from the root
```time go test -i -v -run TestReconcileInstanceWithParameters ./pkg/controller/wip/ -race -c
time ./wip.test -test.v -test.count 10000 -v 10 -logtostderr &> testoutput
```
takes about 3min30sec to run

[5:55]  
https://ibm.ent.box.com/s/tzarns9oungte4ax9p1tetuxoingv5kf should be a link to a folder with a 54meg log file that recreates and has the 467 patch applied.

[5:55]  
@jpeeler @vaikas ^ if you're interested.

vaikas [5:58 PM] 
Awesome, thanks so very much @mbauer that will help me have it down

[5:58]  
Walking the dogs feeding the kids then will tackle this

mbauer [5:58 PM] 
no rush from me, just fooling around with go.

jpeeler [6:03 PM] 
good call on using the race parameter, i forgot that was a thing

mbauer [6:04 PM] 
if it's not valid, turn it off, there are still failures.

[6:04]  
not sure if race is significant to the failure.

vaikas [7:01 PM] 
Well, there are some odd failures that are pretty concerning. If I dump the reason for these errors, I'm seeing that the broker is not sometimes there:
--- FAIL: TestReconcileInstanceWithParameters (0.02s)
    controller_test.go:280: Unexpected condition status: expected True, got False {Type:Ready Status:False Reason:ReferencesNonexistentBroker Message:The instance references a Broker that does not exist}
and sometimes the Service Class is not there:
    controller_test.go:280: Unexpected condition status: expected True, got False {Type:Ready Status:False Reason:ReferencesNonexistentServiceClass Message:The instance references a ServiceClass that does not exist}

Both of these are set up
    sharedInformers.Brokers().Informer().GetStore().Add(getTestBroker())
    sharedInformers.ServiceClasses().Informer().GetStore().Add(getTestServiceClass())

And are used in the controller with:
    serviceClass, err := c.serviceClassLister.Get(instance.Spec.ServiceClassName)
      broker, err := c.brokerLister.Get(serviceClass.BrokerName)

So, I thought that perhaps these methods that create these variables are messed up (since I refactored it), but I replaced getTestBroker/getTestServiceClass with local variables and they still fail. So to my mind it's looking like these listers are flaky, or am I missing something really obvious here?

[7:04]  
Ran with: go test -v -run TestReconcileInstanceWithParameters ./pkg/controller/wip/ -race -count 5000 2>&1 | grep -A 10 -B 10 FAIL > testoutput.new

[7:07]  
If I take out the race flag, things still fail:
vaikas@vaikas-linux3:~/projects/go/src/github.com/kubernetes-incubator/service-catalog$ go test -v -run TestReconcileInstanceWithParameters ./pkg/controller/wip/ -count 5000 2>&1 | grep -A 10 -B 10 FAIL > testoutput.new.norace
vaikas@vaikas-linux3:~/projects/go/src/github.com/kubernetes-incubator/service-catalog$ grep 'got False' testoutput.new.norace     controller_test.go:296: Unexpected condition status: expected True, got False {Type:Ready Status:False Reason:ReferencesNonexistentBroker Message:The instance references a Broker that does not exist}
    controller_test.go:296: Unexpected condition status: expected True, got False {Type:Ready Status:False Reason:ReferencesNonexistentServiceClass Message:The instance references a ServiceClass that does not exist}
    controller_test.go:296: Unexpected condition status: expected True, got False {Type:Ready Status:False Reason:ReferencesNonexistentBroker Message:The instance references a Broker that does not exist}

pmorie [7:31 PM] 
do we know what assertion is causing the test to fail @vaikas

vaikas [7:31 PM] 
Yes the ones on those logs

[7:31]  
Or what do you mean?

pmorie [7:32 PM] 
i mean, can we trace to a specific line of code

[7:32]  
that is causing the test to fail when it decides to flake

vaikas [7:32 PM] 
Yes

pmorie [7:32 PM] 
what's the line

vaikas [7:33 PM] 
Well just went to eat with kids so not in front of my computer but in reconcilenstance the line right after getting a broker

[7:33]  
I'll get back in 30

pmorie [7:33 PM] 
np

vaikas [7:34 PM] 
But it basically does brokerLister.get as listed above and it fails

vaikas [8:12 PM] 
broker, err := c.brokerLister.Get(serviceClass.BrokerName)
    if err != nil {
        glog.Errorf("Instance %v/%v references a non-existent broker %v", instance.Namespace, instance.Name, serviceClass.BrokerName)

[8:13]  
but the error msg is a bit misleading, it might not be missing. I'm changing and rerunning some more tests

vaikas [8:38 PM] 
Ok, last update. I created the following minimalistic repro of what I _think_ I'm seeing:
func TestGetBroker(t *testing.T) {
    _, _, _, _, _, _, sharedInformers, stopCh := newTestController(t)
    defer close(stopCh)

    sharedInformers.Brokers().Informer().GetStore().Add(getTestBroker())
    brokerLister := sharedInformers.Brokers().Lister()

    if _, err := brokerLister.Get("test-broker"); err != nil {
        t.Fatalf("Did not get broker: %v", err)
    }
}

and when I run this:
go test -v -run TestGetBroker ./pkg/controller/wip/ -count 10000 2>&1 | grep -A 10 -B 10 FAIL > testoutput.getbroker.norace.3

I see:
=== RUN   TestGetBroker
E0302 20:35:54.925249    4717 reflector.go:300] github.com/kubernetes-incubator/service-catalog/pkg/client/informers/factory.go:63: Failed to watch *v1alpha1.Binding: unhandled watch: core.WatchActionImpl{ActionImpl:core.ActionImpl{Namespace:"", Verb:"watch", Resource:schema.GroupVersionResource{Group:"servicecatalog.k8s.io", Version:"v1alpha1", Resource:"bindings"}, Subresource:""}, WatchRestrictions:core.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:""}}
E0302 20:35:54.926011    4717 reflector.go:300] github.com/kubernetes-incubator/service-catalog/pkg/client/informers/factory.go:63: Failed to watch *v1alpha1.Broker: unhandled watch: core.WatchActionImpl{ActionImpl:core.ActionImpl{Namespace:"", Verb:"watch", Resource:schema.GroupVersionResource{Group:"servicecatalog.k8s.io", Version:"v1alpha1", Resource:"brokers"}, Subresource:""}, WatchRestrictions:core.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:""}}
E0302 20:35:54.926264    4717 reflector.go:300] github.com/kubernetes-incubator/service-catalog/pkg/client/informers/factory.go:63: Failed to watch *v1alpha1.ServiceClass: unhandled watch: core.WatchActionImpl{ActionImpl:core.ActionImpl{Namespace:"", Verb:"watch", Resource:schema.GroupVersionResource{Group:"servicecatalog.k8s.io", Version:"v1alpha1", Resource:"serviceclasses"}, Subresource:""}, WatchRestrictions:core.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:""}}
--- FAIL: TestGetBroker (0.01s)
       controller_test.go:393: Did not get broker: broker.servicecatalog.k8s.io "test-broker" not found

**Comment by vaikas-google** [comment]
Moving the logs from the chat into the bug so that we don't lose the context.

**Comment by pmorie** [comment]
Closed by #475
